### ISSUE TITLE
Easy way to geocode

### ISSUE THREAD

**Issue by cmongut** [CartoDB/cartoframes#888]
Right now, geocoding a dataset forces a data scientist to learn the SQL API. We can provide a handy method instead like:
Analysis.geocode(dataset, 'street_column_name', ['city_column_name'], ['state_column_name'], ['country_column_name'])

**Comment by andy-esch** [comment]
If possible, it would be great to offer more than just the geocoded address. The bulk geocoding function ([`cdb_bulk_geocode_street_point`](https://carto.com/developers/data-services-api/reference/#geocoding-functions)) offers `precision`, `relevance`, and `match_type`.

We also need to fit in the strategy for geocoding by place names (e.g., City) and postal code.

**Comment by jgoizueta** [comment]
My plan is to implement a direct translation of the current Geocoding analysis using SQL API (internally I'll use `execute_query` from context). So this will work only for stored datasets, we'll discuss possible alternative interfaces later.

I think the interface could look like this:

```
from cartoframes.analysis import GeocodeAnalysis
geocoder = GeocoderAnalysis(credentials)
geocoder.geocode(dataset, street='street', country='country', metadata='metadata')
```

This would wait for completion, we could also have option to run it asynchronously.
The 'metadata' column if specified would contain `relevance` etc in JSON.
We could return some GeocodeResult object with information about the results.

**Comment by jgoizueta** [comment]
This is how the Geocoding analysis looks so far: https://nbviewer.jupyter.org/github/CartoDB/cartoframes/blob/e180a3b4e1288ef651792a74f65ee1ecdbae4c52/examples/06_advanced/geocode.ipynb

In translating the code from the Analysis framework I've found some limitations when using SQL API instead of an actual db connection. I'm using batch queries for the main geocoding processing because it can be time consuming. But this prevent us from using transactions as in the AFW for preventing incorrect result reporting. Also, I'm not sure how this behaves if we run out of credits amid processing, since I cannot commit explicitly in case of error as in AFW (credits are checked before starting the geocoding, but another process could be spending credits from the same account concurrently). 

Apart from this problems (so this is not as robust as with AFW at the moment) it behaves as the AFW Geocoding: a hash column is added to the table to avoid unnecessary geocoding and a lock is used to prevent concurrent geocodings on the same tablel.

**Comment by jgoizueta** [comment]
I've added two new methods to the Geocoder in the PR that can geocode queries and dataframes that are not saved into CARTO. One presents the result as a new table (`geocoded_as_table`) and the other as a dataframe (`geocoded_as_dataframe`). The latter is currently implemented by uploading and geocoding a temporary table, fetching it as a dataframe then deleting it, but that's *implementation detail*; we could avoid the temporary storage (albeit with poorer performance, I'm afraid).

The `geocode` method continues to require a table input, since it geocodes it in place. Here's some minimal documentation: https://github.com/CartoDB/cartoframes/blob/c65ddef15a33c1610e16629dbd1bd4da766dac68/docs/geocode.rst

Summary: the geocoder provides these operations:
* `preview` applicable to any input (table/query/dataframe); doesn't make changes; returns information (currenlty a dictionary)
* `geocode` usable with table dataset which modifies it in place; returns a dictionary as preview
* `geocoded_as_table` can be applied to any input, produces a new table; returns a dictionary that includes a field with the new table as a Dataset.
* `geocoded_as_dataframe` can be applied to any input, produces a dataframe; returns a dictionary that includes a field with the dataframe.

@andy-esch what do you think about this kind of interface for Geocoding in particular, but also having other analyses in mind?

**Comment by cmongut** [comment]
@andy-esch can answer this better than me but here are my 2 cents.
I would only have one method, `geocode`. In other to preview the results, I would pass the option `dry_run=True`. And when the geocoding takes place, I would check inside the method the type of the input and it will return always the result with a dataframe.

**Comment by andy-esch** [comment]
I agree with @cmongut about having a more consolidated interface. At the same time, what's important to me is that people understand that the dry_run/preview option is there because it's pretty powerful and benefits users quite a bit.

If we do decided to consolidate the four class instance methods, I see them as below:
```python
# replace geocoder.preview(ds, street='address', city='city')
geocoder.geocode(ds, street='address', city='city', dry_run=True)

# replace geocoded_as_table
geocoder.geocode(ds, street='address', city='city', table_name='new_table')

# replace geocoded_as_table, deal with duplicate tables
geocoder.geocode(ds, street='address', city='city', table_name='new_table', if_exists='replace'|'fail')

# replace geocoded_as_dataframe, deal with duplicate tables
geocoder.geocode(ds, street='address', city='city')
```

On the return values, I don't know what the best approach is. I could imagine returning a tuple of the result (either a Dataset(tablename) or Dataset(dataframe)) and a dict of the performance of the geocoding: 
```python
(Dataset(...), {'total_rows': 4,'required_quota': 0,'previously_geocoded': 4,'previously_failed': 0,'records_with_geometry': 4})
```
On the metadata, I think it might be helpful to have something about null rows that were not attempted (`'null_rows': 2`) where 'null' is measured on whether all of the input columns are null.

```python
from cartoframes.data.services import Geocoder
from cartoframes.data import Dataset
geocoder = Geocoder()

ds = Dataset('mytable')
geocoded_dataset, geocode_performance = geocoder.geocode(ds, street='address', city='city')
```
---
By the way, I'm curious about what the kwargs interface will look like if a user wants to override the `city='Madrid'` instead of a column name.

**Comment by cmongut** [comment]
@andy-esch why do we have to provide the `if_exists` param when dealing with CARTO tables? I understand that a user can make an error but I think we can live with that.

**Comment by andy-esch** [comment]
I was thinking that might be useful if someone wanted to overwrite their current table with the new one with geometries. But what we should be offering is whether they want to overwrite only `the_geom` column, not the whole table.

**Comment by jgoizueta** [comment]
The `street`, `city`, `state` and `country` arguments all accept either a column name `city="city"` or a literal (using SQL syntax) `city="'Madrid'"` (note the single quotes).

We could use different argument names for those cases:

```
geocoder.geocode(ds, street_column='address', city_column='city')
# vs
geocoder.geocode(ds, street_column='address', city='Madrid')
```

What do you think? Any idea for a more pythonic way?

**Comment by jgoizueta** [comment]
I've been thinking about your proposal for using a single method, and I like it, but I'm missing the *in-place* case (only for table datasets).

Regarding return values, I didn't like to put result dataset/dataframe into the dictionary with the rest of the info, so I like better using a tuple. Even though some cases would only need the dictionary output I'd maintain the (dataset, dict) tuple for all cases.

So we have these cases:

### Dry run (preview):

Use a boolean argument for this (default value False); `table_name` argument would be ignored.

```
geocoder.geocode(ds, street='address', city='city', dry_run=True)
```

Returns `(ds, {'required_quota':...})` Where `ds` is just the input dataset.

### Table

Geocode a table *in place* (i.e. fill up `the_geom`/`metadata`, add hash column). Input dataset must be a table or it will fail:

```
geocoder.geocode(ds, street='address', city='city', dry_run=True)
```

Returns `(ds, {'required_quota':...})` Where `ds` is just the updated input dataset.

### New table

Put results in a new table, by given its name in a `table_name` argument. This could have the `if_exists` option too as mentioned.

```
geocoder.geocode(ds, street='address', city='city', table_name='new_table')
```

Returns `(new_ds, {'required_quota':...})` Where `new_ds` is a table Dataset for the new table.

### Dataframe

Puts the result in a new dataframe.
This is the case that I'm not sure how to differentiate from in-place geocoding. I'm using here a `dataframe` flag argument but I'm not too happy with it:

```
geocoder.geocode(ds, street='address', city='city', dataframe='new_table')
```

Returns `(new_ds, {'required_quota':...})` Where `new_ds` is a Dataframe Dataset.

**Comment by jgoizueta** [comment]
Some notes on comparing this geocoder with [GeoPy](https://geopy.readthedocs.io):

GeoPy provides a method `geocode(self, query, exactly_one=True, timeout=DEFAULT_SENTINEL)` which operates on a single address. The address can be provided as a simple text string or as a dictionary with separate fields, but in that case the fields available and its names vary per provider. It also admits provider-specific additioinal arguments (e.g. bbox for Here). 

It also provides reverse geocoding for some providers.

It can obtain not only a single result (`exactly_one`) but a list of them.

To allow applying it to a Dataframe, since it will make a request per row, a [`RateLimiter`](https://github.com/geopy/geopy/blob/472089da824d6b61321f1873a60514fb9367fc26/geopy/extra/rate_limiter.py) class is provided.

On the other hand, our Geocoder  (based on CARTO data services) has a unified interface to all providers (allowing structured input) and geocodes multiple addresses in a single db query, where internal functions execute it in batches and take care of rate limits, etc. This makes it potentially more efficient (specially if the data is already in the db).

In summary: GeoPy provides access to provider-specific features, including reverse geocoding and multiple results. It makes a request for each individual address and provides a convenient way to apply it to a dataframe. We provide metered *least-common* access to configurable providers in an efficient way for db tables.

**Comment by jgoizueta** [comment]
After suggestion by @cmongut I think we can simplify my previous proposal and avoid needing the `dataframe` parameter:

### Dry run (preview):

Works with any kind of dataset (table, query or dataframe)
Use a boolean argument for this (default value False); `table_name` argument would be ignored.

```
geocoder.geocode(ds, street='address', city='city', dry_run=True)
```

Returns `(ds, {'required_quota':...})` Where `ds` is just the input dataset.

### Create New table for the results

Works for any kind of dataset (table, query or dataframe).
Put results in a new table, by given its name in a `table_name` argument. This could have the `if_exists` option too as mentioned.

```
geocoder.geocode(ds, street='address', city='city', table_name='new_table')
```

Returns `(new_ds, {'required_quota':...})` Where `new_ds` is a table Dataset for the new table.

### Don't create new tables

```
geocoder.geocode(ds, street='address', city='city')
```

If the input dataset is a table, then it is modified in place (i.e. fill up `the_geom`/`metadata`, add hash column).
Returns `(ds, {'required_quota':...})` Where `ds` is the updated input dataset.

If the input is a dataframe or a query, then a new dataframe is created with the results
Returns `(new_ds, {'required_quota':...})` Where `new_ds` is a newly created Dataframe Dataset.

**Comment by jgoizueta** [comment]
I have one doubt after talking with @cmongut 

My implementation expects always a Dataset object, even if it's a dataframe not to be uploaded to CARTO. And it always returns the results as a Dataset (for newly created dataframes too).

Do you think take Dataframes directly (and/or return dataframes without wrapping them in a Dataset)?

**Comment by cmongut** [comment]
@alejandrohall could you take a look to this?

**Comment by alejandrohall** [comment]
Regarding the definition, seems to be good for me.

Regarding the data type input for the geocoder, I think could make sense to allow geocode a dataframe itself (without being a Dataset). If we allow geocoding a dataframe I think we should return a dataframe too. Same for Dataset, return a Dataset if the input is a Dataset.

**Comment by jgoizueta** [comment]
I'm changing the implementation to accept DataFrames then. I have one doubt about a particular case though, @alejandrohall :

When the input is a DataFrame, but the optional argument `table_name` is used, so that the results will be stored in a newly created table. should the return value be a DataFrame containing the geocoded results or a Dataset which represents the new table?

**Comment by alasarr** [comment]
For me, it looks great!

A great improvement will to work on the quality of the geocoding. 

I suggest:
-  We could return a column with precision/match type. It will allow checking how good was the geocoding. 
-  Preview methods could return some kind of `describe` of the datasets saying the % of geocoding operations with good accuracy.

I've thought about a parameter to return only addresses with good matches/precision (for example I don't want to get the errors with geometries at the center of the city or province) but maybe the user can easily do it via Pandas. Thoughts about this?

**Comment by cmongut** [comment]
I think it makes sense but I would try to merge a version with master without that.
Once we do that, we can iterate over it and add that :)

**Comment by alasarr** [comment]
Totally agree @cmongut

**Comment by alejandrohall** [comment]
> When the input is a DataFrame, but the optional argument `table_name` is used, so that the results will be stored in a newly created table. should the return value be a DataFrame containing the geocoded results or a Dataset which represents the new table?

@jgoizueta  I would return a DataFrame because it is confusing for the user if you don't use the argument tablename to receive a DataFrame but if you use it to receive a Dataset, IMHO.

**Comment by jgoizueta** [comment]
> * We could return a column with precision/match type. It will allow checking how good was the geocoding.

Regarding the precision of the geocoding. We can request per row metadata which includes `precision`, `relevance` and `match_types` in a JSON object.

From https://carto.com/developers/data-services-api/reference/ the fields in this object are:

Name | Type | Description
-- | -- | --
precision | text | One of precise or interpolated.
relevance | number | Relevance factor, from 0 to 1, higher being more relevant.
match_type | text | Array with one of point_of_interest, country, state, county, locality, district, street, intersection, street_number, postal_code. Empty array if match type is unknown.

 I've just modified this analysis so that the metadata column is created if it doesn't exist (if it is requested) which is convenient for geocoding dataframes or queries.

I'm not sure if we should provide separate columns instead of a JSON for convenience; currenly we can do this:

```python
from cartoframes.auth import set_default_credentials, Credentials
credentials = Credentials.from_file()
set_default_credentials(credentials)
from cartoframes.data import Dataset
from cartoframes.analysis import GeocodeAnalysis
import json
import pandas as pd

dataset = Dataset('carto_places')
gc = GeocodeAnalysis()
ds, info = gc.geocode(dataset, street='address', country="country", metadata='gcmeta')
df = ds.download()
# extract metadata into separate columns:
df = pd.concat([df.reset_index().drop('gcmeta', axis=1), pd.io.json.json_normalize(df.gcmeta.apply(json.loads))], axis=1)
print(df)
```

```
   cartodb_id          place                               address country                carto_geocode_hash                                           the_geom precision  relevance      match_types
0           1       CARTO NY     114 W 26th St, New York, NY 10001     USA  333cb1971579e7908f92d439a95d3ae3  0101000020E610000028B85851837F52C025404D2D5B5F...   precise       0.99  [street_number]
1           2   CARTO Madrid             Gran Via 46, 28013 Madrid   Spain  41acbd170c4dc9175e9cdda5a66c4351  0101000020E610000036B05582C5A10DC0A032FE7DC635...   precise       0.86         [street]
2           3       CARTO DC  4075 Wilson Blvd, Arlington VA 22203     USA  1e998c791d577302df42cefc40b03c74  0101000020E610000037A6272CF14653C031D3F6AFAC70...   precise       0.95  [street_number]
3           4  CARTO Sevilla           Calle Ebro 1, 41012 Sevilla   Spain  17c115e4b766018f635a9ea980da3783  0101000020E6100000912C6002B7EE17C0C45A7C0A80AD...   precise       0.91  [street_number]
```

**Comment by jgoizueta** [comment]
PR here: #943 No tests yet, but it could use some CR

**Comment by jgoizueta** [comment]
I'm moving the geocoding analysis from the `analysis` namespace to `data.services` as discussed with @oleurud and @alasarr 
The idea is that the CARTO LDS Geocoding & Isocrones are part of CARTO data services.

I've considered renaming the class `GeocodeAnalysis` to `GeocodeService`, but I think I'll go with just `Geocode` (that's what name spaces are for, aren't they?)

**Comment by oleurud** [comment]
+1 to Geocode

**Comment by oleurud** [comment]
About your comment in the PR

> If we drop the direct DataFrame support we would just require users to use Dataset(dataframe) to pass dataframes and dataset.dataframe to fetch them from the results. Is it really important/convenient for users to be able to use DataFrames directly?

I think we should go with both approaches. I am at the very beginning of the review, so I dont know the code and implementation details but is not enough using the `isinstance` method to check if the data is a Dataset or a DataFrame?

Anyway, it is a new confirmation we need to tackle this problem https://github.com/CartoDB/cartoframes/issues/910 @cmongut @alrocar

**Comment by andy-esch** [comment]
I'm 100% on board with the API around Geocode, except for the custom literal string, like in the country name:

```python
geocoded_dataframe, info = gc.geocode(
    dataframe,
    street='address',
    city='city',
    country="'Spain'"
)
```

I want to explore some alternatives to achieve this:

1. Pass it as a string like the other arguments (all without quotes), but do a check to see if the column exists as that name. Assume it's a country, if not, proceed. This is kinda messy and too magical / prone to failure or confusing results.
2. Pass as `dict` and string literal (overload types):
    * `country={'literal': 'Spain'}` for string literal,
    * `country={'column': 'country_col'}` for column
    * `country='country_col'` for column names.
3. Custom class for custom name? Custom class for columns, similar to the `Column` class that's already in `Dataset`
  * `country=Column('country_col')`
  * `country='Spain'`

Honestly, I don't feel satisfied with any of these, but feel like the second option is my favorite even though it's quite a bit of typing and would require documentation lookups.

**Comment by jgoizueta** [comment]
Hey @andy-esch I also wasn't quite happy with the literal strings solution, so I'll be happy to change it to something we agree on even if not a perfect solution.

I could live with either number 2 or 3, I'm not sure which one I prefer.

Have you also considered the alternative I [proposed above](https://github.com/CartoDB/cartoframes/issues/888#issuecomment-528269697)? i.e. having a pair of parameters for each attribute, like `country_column` and `country`, or to be more explicit `country_column_name` and `country_literal` or perhaps `country_name`:
* `country_column='country_col'` for column
* `country_literal='Spain'` for string literal

After I little more thought, I think I like number 3 personally, but I think number 2 would be less prone to user mistakes (passing a column name which will be interpreted as a literal, and will cause no error, but poor geocoding results).

I think that number 2, with the simplified default form for column names `country='country_col'` could be the best compromise. So, since it's @andy-esch's favourite, unless someone has any other idea I will go with it.

Now I'm not sure what could be most clear for users for the `literal` option:
* `country={'literal': 'Spain'}` Seems natural to me because we're using a literal value, a specific country, but will it be natural for users? (as a programmer I'm used to the concept of a literal value or string)
* `country={'fixed': 'Spain'}` I think an alternative could be to use a term (I don't quite like `fixed`) to convey the idea that we define a common value for the country for all the data instance, unlike when we use a column name which defines an individual value for each instance.

**Comment by jgoizueta** [comment]
@malgar has made a suggestion I like for the `literal`/`fixed`alternative: `value`.

To me it's a clear distinction between specifying a value or a column for the attribute:

    * `country={'value': 'Spain'}` for fixed value (string literal)
    * `country={'column': 'country_col'}` for column
    * `country='country_col'` shortcut for column names.

**Comment by andy-esch** [comment]
Thanks for thinking through it with me @jgoizueta! Big thumbs up for `value`.

I guess if we go the way of having `dict` we should do some input validation to make sure the user puts in one of the two valid keys. If the user puts in a column name that is not valid, I'm guessing the geocoder will throw an error before running with the query is being planned? If that's the case, we get a check for a valid column name for free.

**Comment by jgoizueta** [comment]
Having seen some tests @cmongut is doing, there's one thing that worries me about the geocoding metadata.

Since the underlying LDS functions don't use structured geocoding services (we just append city, state to the address separated by commas), the metadata is very unreliable (at least for TomTom): we find results out of the specified city/state with high scores; I guess because the city/state are interpreted as part of the street address.

As I user, having provided city, state as separate parameters I would expect to have them into account, and I would consider points with a high precision and relevance in the metadata to be highly reliable matches, but (again, at least for TomTom) they may as well have been *precisely* and *relevantly* matched to addresses in a different city or state.

**Comment by alrocar** [comment]
👏
